Automattic/jetpack
# Enable Support to perform utf8 conversions during checksum calculations

The Sync Checksum/Fix routine calculates a hash of remote data to be compared against the WP.com cache. The WP.com cache uses latin1 charset which causes differences in the calculated checksums if remote db uses utf8 or other charsets. To resolve the issue we need to convert text based fields to utf8 so that we have consistent checksums. 

To minimize performance impacts we are requiring that the requests for the checksums contain  the perform_text_conversion argument to trigger the conversions. 

This change will not cause any impact with the release itself but instead will be controlled by toggling the request from the WP.com side so we can monitor behavior and disable if it causes degradation of performance for the queries or job run times.

Fixes # https://wp.me/p9dueE-3gM

#### Changes proposed in this Pull Request:
* Enable support to perform utf8 conversions on text fields when generation checksums.

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Ensure your test site has db charset of utf8
* Add meta_value entries to you post content that encludes encoded characters. ex: Colleen Oâ€™Brien
* Have you test site use the current version of Jetpack (master branch is fine)
* From a WP.com sandbox run the following commands
* wp shell --url=XXXX (test site)
* $validator = new Jetpack_Sync_Validator( get_current_blog_id() );
* $checksum_results = $validator->validate_checksum(null, false, true);
* record output for later comparison
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], false);
* The meta_key which has the encoded characters should be identified as different
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], true);
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], false);
* The meta_key which has the encoded characters should still be identified as different (could not be fixed)
* Update test site to this branch/pr
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], false);
* The meta_key which has the encoded characters should be identified as different
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], true);
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], false);
* The meta_key which has the encoded characters should still be identified as different (could not be fixed)
* Apply Diff XXXXXX to sandbox
* wp shell --url=XXXX (test site)
* $validator = new Jetpack_Sync_Validator( get_current_blog_id() );
* $checksum_results = $validator->perform_cache_site_audit(['postmeta'], false, true);
* The meta_key which has the encoded characters should not be identified as different
* $checksum_results = $validator->validate_checksum(null, false, true, false);
* postmeta is identified with differences
* all other table behavior matches pre-patch behavior.
* $checksum_results = $validator->validate_checksum(null, false, true, true);
* postmeta is identified with no differences
* all other table behavior matches pre-patch behavior if they don't have encoding in fields (comment meta, terms, term_taxonomy, postmeta)
